can/bus.py

```diff
@@ -2,79 +2,174 @@
+    Concrete implementations *have to* implement the following:
+        * :meth:`~can.BusABC.send` to send individual messages
+        * :meth:`~can.BusABC._recv_internal` to receive individual messages
+          (see note below)
+        * set the :attr:`~can.BusABC.channel_info` attribute to a string describing
+          the underlying bus and/or channel
+
+    They *might* implement the following:
+        * :meth:`~can.BusABC.flush_tx_buffer` to allow discrading any
+          messages yet to be sent
+        * :meth:`~can.BusABC.shutdown` to override how the bus should
+          shut down
+        * :meth:`~can.BusABC.send_periodic` to override the software based
+          periodic sending and push it down to the kernel or hardware
+        * :meth:`~can.BusABC._apply_filters` to apply efficient filters
+          to lower level systems like the OS kernel or hardware
+        * :meth:`~can.BusABC._detect_available_configs` to allow the interface
+          to report which configurations are currently available for new
+          connections
+        * :meth:`~can.BusABC.state` property to allow reading and/or changing
+          the bus state
+
+    .. note::
+
+       Previously, concrete bus classes had to override :meth:`~can.BusABC.recv`
+       directly instead of :meth:`~can.BusABC._recv_internal`, but that has
+       changed to allow the abstract base class to handle in-software message
+       filtering as a fallback. Older (custom) interfaces might still be
+       implemented like that and thus might not provide message filtering.
 
     """
 
-    #: a string describing the underlying bus channel
+    #: a string describing the underlying bus and/or channel
     channel_info = 'unknown'
 
     @abstractmethod
     def __init__(self, channel=None, can_filters=None, **config):
-        """
+        """Construct and open a CAN bus instance of the specified type.
+
+        Subclasses should call though this one with all parameters as
+        it applies filters (for now).
+
         :param channel:
             The can interface identifier. Expected type is backend dependent.
 
         :param list can_filters:
-            A list of dictionaries each containing a "can_id", a "can_mask",
-            and an "extended" key.
-
-            >>> [{"can_id": 0x11, "can_mask": 0x21, "extended": False}]
-
-            A filter matches, when ``<received_can_id> & can_mask == can_id & can_mask``
+            See :meth:`~can.BusABC.set_filters` for details.
 
         :param dict config:
             Any backend dependent configurations are passed in this dictionary
         """
-        pass
+        self.set_filters(can_filters)
 
     def __str__(self):
         return self.channel_info
 
-    @abstractmethod
     def recv(self, timeout=None):
         """Block waiting for a message from the Bus.
 
-        :param float timeout: Seconds to wait for a message.
+        If the concrete bus does not override it, this method makes sure
```

Comment:
This is reading like library developer documentation - not user documentation.

This class along with the `Message` data class is essentially the primary interface to python-can. The front page of the docs links directly to [API Docs -> Bus](https://python-can.readthedocs.io/en/develop/bus.html#bus). I think this needs to be moved into an interface developers section of the docs.